### DISCUSSION TITLE
Indicator variable for restricting feature

### DISCUSSION THREAD

**Discussion by khl02007** [flatironinstitute/nemos#473] [Q&A] [answered]
I'm trying to estimate place fields with GLM via nemos. The behavior consists of four different trajectory types, and I want to estimate the place field separately for different trajectory types. I want to do this by multiplying the basis functions for 2d position by an indicator variable. I'm currently using `nmo.basis.IdentityEval` for that but I'm not sure if this is the right approach. Here is more detail about how I'm setting up the model. `position_2d` is a `TsdFrame` of `x` and `y` position and `in_center_to_left` etc is a `Tsd` of 1s and 0s (1 if time point is during the trajectory type; 0 otherwise). Is this the right way to do this?

```python
position_2d_basis_center_to_left = nmo.basis.BSplineEval(
    n_basis_funcs=15, label="center_to_left_x"
) * nmo.basis.BSplineEval(n_basis_funcs=15, label="center_to_left_y")
position_2d_basis_center_to_left.label = "position_center_to_left"

position_2d_basis_center_to_right = nmo.basis.BSplineEval(
    n_basis_funcs=15, label="center_to_right_x"
) * nmo.basis.BSplineEval(n_basis_funcs=15, label="center_to_right_y")
position_2d_basis_center_to_right.label = "position_center_to_right"

position_2d_basis_left_to_center = nmo.basis.BSplineEval(
    n_basis_funcs=15, label="left_to_center_x"
) * nmo.basis.BSplineEval(n_basis_funcs=15, label="left_to_center_y")
position_2d_basis_left_to_center.label = "position_left_to_center"

position_2d_basis_right_to_center = nmo.basis.BSplineEval(
    n_basis_funcs=15, label="right_to_center_x"
) * nmo.basis.BSplineEval(n_basis_funcs=15, label="right_to_center_y")
position_2d_basis_right_to_center.label = "position_right_to_center"

path_identity_center_to_left = nmo.basis.IdentityEval(label="center_to_left")
path_identity_center_to_right = nmo.basis.IdentityEval(label="center_to_right")
path_identity_left_to_center = nmo.basis.IdentityEval(label="left_to_center")
path_identity_right_to_center = nmo.basis.IdentityEval(label="right_to_center")

basis = (position_2d_basis_center_to_left*path_identity_center_to_left
    + position_2d_basis_center_to_right*path_identity_center_to_right
    + position_2d_basis_left_to_center*path_identity_left_to_center
    + position_2d_basis_right_to_center*path_identity_right_to_center
    )

X = basis.compute_features(position_2d['x'], position_2d['y'], in_center_to_left,
                           position_2d['x'], position_2d['y'], in_center_to_right,
                           position_2d['x'], position_2d['y'], in_left_to_center,
                           position_2d['x'], position_2d['y'], in_right_to_center,)

glm = nmo.glm.GLM(
    solver_kwargs=dict(tol=10**-12),
    solver_name="LBFGS"
)
glm.fit(X, count_epoch[:,0])
```

**Comment by sjvenditto** [comment]
Hello! If I'm understanding your data correctly, it looks like all the trajectory types are independent. In that case, why not fit four separate models to estimate the place fields for each trajectory? You could use pynapple functions to restrict your data to each trajectory to do this. Is there a specific reason you want to fit them all jointly?

**Reply by khl02007** [reply]
Thanks. Yes, that's a good point. I wanted to fit the four different place fields jointly (one for each trajectory) because I was also planning to add other (motoric) variables to this model and those are not trajectory specific; i.e. I would like to use the entire data to estimate their contributions rather than separately for each trajectory type. I also wasn't sure if `nmo.basis.IdentityEval` really works as I think it does since I couldn't find much about it in the documentation.

**Reply by sjvenditto** [reply, marked as answer]
I that case, yes you can use `IdentityEval` in this way. All this basis function does is give back the input scaled between 0 and 1 (or scaled differently if you provide the `bounds` parameter), so for a boolean value it will be unchanged. If you want to use NeMoS to combine basis functions, this is probably the best way to do this. Something to keep in mind, if you want to compare the trajectories within a single model, it will be important to have the same number of time points/observations for each trajectory, otherwise the results may be biased towards trajectories with more observations

**Reply by khl02007** [reply]
Got it, thank you. That's a good point about the relative number of observations for each trajectory. I think they are unequal, though I think I have enough for each trajectory type such that the GLM coefficients / place fields can be estimated accurately. But maybe this is more of a problem if I use regularization?

**Reply by BalzaniEdoardo** [reply]
> [!NOTE]
  > Approach 2 requires installing NeMoS from the development branch. In the current release, `CustomBasis` with multiple inputs has a validation bug. Install from source with:
  > ```
  > pip install git+https://github.com/flatironinstitute/nemos.git@development
  > ```


If I understand your implementation correctly, there's a more efficient approach. Your current method creates K separate design matrices for position and uses IdentityEval to mask each category — this works but computes the B-spline basis K times.

Instead, you can use CustomBasis to compute the position basis once and then broadcast it across categories in a single pass using einsum. See the attached script for a complete example. 



```python
"""
Example: Efficient Categorical Design Matrices with CustomBasis

This script demonstrates two approaches to creating design matrices that
interact spatial position with categorical variables (e.g., trajectory type).

Approach 1 (Current): Creates K separate design matrices and uses IdentityEval
                      to mask each category. Simple but computes bases K times.

Approach 2 (Efficient): Uses CustomBasis to compute the position basis once,
                        then broadcasts across categories via einsum.
"""

import jax
import numpy as np
import nemos as nmo
from functools import partial


# =============================================================================
# Generate sample data
# =============================================================================

# 2D position grid (100 samples)
x, y = np.meshgrid(np.linspace(0, 2 * np.pi, 10), np.linspace(0, 2 * np.pi, 10))
x = x.flatten()
y = y.flatten()

# Categorical variable: 3 trajectory types with different sample counts
n_categories = 3
trajectory_index = np.array([0] * 30 + [1] * 20 + [2] * 50)


# =============================================================================
# Approach 1: Your current method
# Creates one basis per category, each requiring x, y, and a boolean mask
# =============================================================================

b1 = ((nmo.basis.BSplineEval(5) ** 2) * nmo.basis.IdentityEval()).set_params(label="traj 1")
b2 = ((nmo.basis.BSplineEval(5) ** 2) * nmo.basis.IdentityEval()).set_params(label="traj 2")
b3 = ((nmo.basis.BSplineEval(5) ** 2) * nmo.basis.IdentityEval()).set_params(label="traj 3")

bas_current = b1 + b2 + b3

# Each basis gets (x, y, mask) -> 9 inputs total
X_current = bas_current.compute_features(
    x, y, trajectory_index == 0,
    x, y, trajectory_index == 1,
    x, y, trajectory_index == 2
)


# =============================================================================
# Approach 2: Efficient implementation using CustomBasis
# Computes position basis once, then broadcasts to all categories
# =============================================================================

def categorical_basis_func(x, y, traj_index, bas):
    """
    Compute design matrix for position x category interaction.

    Parameters
    ----------
    x, y : array-like
        Position coordinates.
    traj_index : array-like
        Category index.
    bas : Basis
        The position basis to use.

    Returns
    -------
    X : ndarray, shape (n_samples, n_basis_funcs * n_categories)
        Design matrix where each category gets its own copy of the basis,
        zeroed out for samples not in that category.
    """
    # Compute position basis ONCE for all samples
    X = bas.compute_features(x, y)  # shape: (n_samples, n_basis_funcs)

    # One-hot encode categories
    traj_one_hot = jax.nn.one_hot(traj_index, n_categories)  # shape: (n_samples, n_categories)

    # Broadcast: multiply each basis function by each category indicator
    # "tk, tc -> tck" means: for each sample t, basis k, category c
    # Result: X[t,k] * one_hot[t,c] for all combinations
    X_categorical = jax.numpy.einsum("tk, tc -> tck", X, traj_one_hot)

    # Flatten to (n_samples, n_basis_funcs * n_categories)
    return X_categorical.reshape(X.shape[0], -1)


# Define the position basis (computed only once internally)
bas_position = nmo.basis.BSplineEval(5) ** 2
bas_position.label = "position"

# Create partial function with fixed parameters

func = partial(
    categorical_basis_func,
    bas=bas_position
)

# Wrap in CustomBasis (can be combined with other bases using +, *, etc.)
bas_efficient = nmo.basis.CustomBasis(func)
X_efficient = bas_efficient.compute_features(x, y, trajectory_index)


# =============================================================================
# Verify equivalence
# =============================================================================

print("Max abs difference (current vs efficient):", np.max(np.abs(X_current - X_efficient)))


# =============================================================================
# Extra: Reshape coefficients by category after fitting
# =============================================================================

# Set input shape to enable split_by_feature (3 categories, each with 5x5=25 basis funcs)
bas_position.set_input_shape(n_categories, n_categories)


# Example: fit GLM and reshape coefficients
w_true = np.random.randn(n_categories * 25)
counts = np.random.poisson(np.exp(X_efficient.dot(w_true) - 1))

model = nmo.glm.GLM(solver_name="LBFGS").fit(X_efficient, counts)

# Reshape fitted coefficients -> shape: (n_categories, n_basis_x * n_basis_y)
coef_by_category = bas_position.split_by_feature(model.coef_, axis=0)["position"]
print("Coefficient shape by category:", coef_by_category.shape)

```

**Reply by khl02007** [reply]
@BalzaniEdoardo Thank you for providing this. This does seem much more efficient. Do you think this could be a new basis class that nemos provides? It could take as input a baseline basis (e.g.`bas_position` in the example) and an indicator variable (`trajectory_index` in the example) and give as output the baseline basis for each value of the indicator variable. If this kind of design comes up often this may be useful to have since `einsum` and `partial` are somewhat esoteric (to me anyway).

**Reply by BalzaniEdoardo** [reply]
That's a good point, and it makes sense to have a basis like that. If you have suggestions for a good name let me know. I'll open an issue about this

**Reply by khl02007** [reply]
Thanks. Here are some names that might be appropriate (feel free to not use them if you don't like them!): `CategoricalBasis`, `GatedBasis`, `CategoricalGatedBasis`, `CategoricalMaskedBasis`, `CategoryExpandedBasis`.
Possible docstring: Replicate a base basis across categories and mask by a categorical variable; features are “on” only when the category is active.

**Comment by BalzaniEdoardo** [comment]
Also, note that I used the internal basis (`bas_position`) to split the coefficients because if I used the CustomBasis,  because a product of 2 bases anymore, but as a single basis with 75 elements. So split by feature will give you:

```ipython
>>> bas_efficient.split_by_feature(model.coef_, axis=0)["CustomBasis"].shape
(75, 1)
```

**Comment by khl02007** [comment]
@sjvenditto @BalzaniEdoardo By the way, I was trying to fit the full model of spatial bins for the four trajectories + motor variables:
```python
# basis
position_2d_basis_center_to_left = nmo.basis.BSplineEval(
    n_basis_funcs=10, label="center_to_left_x"
) * nmo.basis.BSplineEval(n_basis_funcs=10, label="center_to_left_y")
position_2d_basis_center_to_left.label = "position_center_to_left"

position_2d_basis_center_to_right = nmo.basis.BSplineEval(
    n_basis_funcs=10, label="center_to_right_x"
) * nmo.basis.BSplineEval(n_basis_funcs=10, label="center_to_right_y")
position_2d_basis_center_to_right.label = "position_center_to_right"

position_2d_basis_left_to_center = nmo.basis.BSplineEval(
    n_basis_funcs=10, label="left_to_center_x"
) * nmo.basis.BSplineEval(n_basis_funcs=10, label="left_to_center_y")
position_2d_basis_left_to_center.label = "position_left_to_center"

position_2d_basis_right_to_center = nmo.basis.BSplineEval(
    n_basis_funcs=10, label="right_to_center_x"
) * nmo.basis.BSplineEval(n_basis_funcs=10, label="right_to_center_y")
position_2d_basis_right_to_center.label = "position_right_to_center"

path_identity_center_to_left = nmo.basis.IdentityEval(label="center_to_left")
path_identity_center_to_right = nmo.basis.IdentityEval(label="center_to_right")
path_identity_left_to_center = nmo.basis.IdentityEval(label="left_to_center")
path_identity_right_to_center = nmo.basis.IdentityEval(label="right_to_center")

speed_basis = nmo.basis.MSplineEval(n_basis_funcs=10, label="speed")
acc_basis = nmo.basis.MSplineEval(n_basis_funcs=10, label="acc")
headdir_basis = nmo.basis.MSplineEval(n_basis_funcs=10, label="headdir")
headdir_vel_basis = nmo.basis.MSplineEval(n_basis_funcs=10, label="headdir_vel")
headdir_speed_basis = nmo.basis.MSplineEval(n_basis_funcs=10, label="headdir_speed")

# models
models = (
        position_2d_basis_center_to_left * path_identity_center_to_left
        + position_2d_basis_center_to_right * path_identity_center_to_right
        + position_2d_basis_left_to_center * path_identity_left_to_center
        + position_2d_basis_right_to_center * path_identity_right_to_center
        + speed_basis
        + acc_basis
        + headdir_basis
        + headdir_vel_basis
        + headdir_speed_basis
)
```
and have found that while I can fit single neurons well with `nemos.glm.GLM`, fits with `nemos.glm.PopulationGLM` are much worse. This was true when the same regularization was applied to both. I understand that `nemos.glm.PopulationGLM` solves a different optimization problem that takes into account all the neurons at once (rather than being just a parallelized version of `nemos.glm.GLM`) but I wasn't really sure what to do next or how to diagnose things. Any suggestions would be appreciated.

**Reply by BalzaniEdoardo** [reply]
Yes, this is a known issue. In some cases, the GLM fitting procedure may stop early and return sub-optimal coefficients. We have discussed exposing an explicit *fitting modality* for population GLMs. The default would remain the current behavior—jointly fitting all neurons by maximizing the sum of their log-likelihoods. An alternative would be to `vmap` over the neural axis (i.e., fit one GLM per neuron). The latter is equivalent to fitting *N* single-neuron GLMs, but it is significantly slower, especially without a GPU.

What I recommend checking to debug this issue is the following.

### 1. Inspect the solver state

Check `model.solver_state_`. Depending on the solver, this should contain a field such as `model.solver_state_.iter_num`, which indicates how many optimization iterations were executed.

The `solver_kwargs["maxiter"]` parameter caps this value (the default is 500).

* If `iter_num` is **much smaller than 500**, the solver likely stopped early. In this case, you may want to **decrease the tolerance** to reduce the risk of premature stopping.
* If `iter_num` is **equal to 500**, the solver reached the iteration cap without converging. In this case, you should **increase `maxiter`**.

Both parameters can be adjusted, for example:

```python
solver_kwargs = {"maxiter": 2000, "tol": 1e-12}
```

Use values that are appropriate for your data and model scale.

### 2. Enable float64 precision

In general, enabling double precision improves numerical stability for GLM optimization. You can do this at the beginning of your script:

```python
import jax
jax.config.update("jax_enable_x64", True)
```

### 3. Use the LBFGS solver

Make sure you are using `solver_name="LBFGS"`. In my experience, this is the most reliable solver we currently provide.

---

If none of the above resolves the issue, there is an additional workaround that is not yet exposed in the public API. It consists of explicitly performing a `vmap` over single-neuron fits, using the population fit as a warm start.

```python
import jax
from nemos.glm.params import GLMParams
import nemos as nmo
import numpy as np

# generate some data
n_features, n_neurons = 3, 2
X = np.random.randn(100, n_features)
w = np.random.randn(n_features, n_neurons)
i = -np.ones(n_neurons)
y = np.random.poisson(np.exp(X.dot(w) + i))

# fit a population model (used for warm start)
model = nmo.glm.PopulationGLM()
model.fit(X, y)

# wrap the (potentially suboptimal) parameters in the internal container
init_params = GLMParams(model.coef_, model.intercept_)

# define a single-neuron GLM
model_single = nmo.glm.GLM()

# initialize the solver and its state
model_single.initialize_solver_and_state(
    X,
    y[:, 0],
    model_single.initialize_params(X, y[:, 0])
)

# vectorize over the neuron axis
vmap_solver_run = jax.vmap(
    model_single.solver_run,
    in_axes=(GLMParams(1, 0), None, 1),
    out_axes=-1,
)

pars, state, _ = vmap_solver_run(init_params, X, y)

# replace coefficients, intercepts, and solver state
model.coef_ = pars.coef
model.intercept_ = pars.intercept
model.solver_state_ = state
```

Note that the names `solver_run` and `initialize_solver_and_state` are likely to change, so this code may break in the future. Once we implement the `vmap` fitting mode internally, the workflow will be much simpler and should look roughly like this:

```python
model = nmo.glm.PopulationGLM()
model.fit(X, y)

init_params = model.coef_, model.intercept_
model.fit_type = "vmap"
model.fit(X, y, init_params)
```

**Reply by khl02007** [reply]
Thank you, that helped. I will just note that other things that helped were rescaling the features (e.g. zscoring) and also getting rid of neurons with very low firing rate (in y).
